Refactor

dselans opened the pull request:
Some potential refactor work for dkenv. Will add tests shortly.

Don't merge - let's discuss first when we have a moment!

@mikkergimenez @idleyoungman

Commit:
Refactor

- moved majority of code away from `main` and into their own packages
- replaced viper with kingpin lib; hardened and standardized cli parsing
- replaced log usage with logrus; standardized output
- simplified main()
- replaced os.Current() (cgo-only) with mitchellh's homedir lib (avoids homedir lookup issues on various OS's)
- enabled debug mode
- download attempts now check status code (previously, we would download whatever a 404 returned (ie. a .png or .html)
- no longer saving 'docker' to ~/.dkenv/ - instead, save specific docker version to ~/.dkenv/docker-x.x.x and symlink to bin dir (/usr/local/bin by default)
- still need to add tests and godep

dselans commented on dkenv/dkenv.go:
```diff
@@ -0,0 +1,206 @@
+//     - create symlink
+//
+// TODO This will be a pain to test - refactor when possible.
+func (d *Dkenv) UpdateSymlink(version string) error {
```
This is really meh.. would like to update this prior to merge; it'll also make testing it _much_ simpler.

idleyoungman committed:
Rename dkenv package to lib

This avoids a conflict when building, as the binary is named dkenv.

idleyoungman committed:
Ignore built binary

idleyoungman commented on lib/downloader.go:
```diff
@@ -49,102 +31,81 @@ type PassThru struct {
-
-	if req.URL.Scheme != "https" {
-		return Err_HTTP_Redirect
+	if err := ioutil.WriteFile(d.DkenvDir+"/docker-"+version, body, 0777); err != nil {
```
What about 0755 for FileMode?

dselans replied:
Agreed - bad copy/paste.

idleyoungman committed:
Change mode of binary to 0755

idleyoungman commented:
:joy_cat: LGTM!

idleyoungman committed:
Merge branch 'master' into dselans/refactor

idleyoungman committed:
Fix error naming

idleyoungman commented:
Fixed the merge conflicts while I was at it.

Commit:
tests

- added cli tests
- stubbed out lib_tests*

Note: Not sure about the easiest way to mock a failing os.Stat(), so skipping those cases in cli package.

Commit:
godeps

dselans commented:
Added (some) tests, godeps - this should be in a good enough state at this point. Waiting on @mikkergimenez @idleyoungman for review.

Will flesh out the rest of the tests when I have a bit more time.

```
dselans:dkenv dselans$ godep-osx go test ./... -cover
?       github.com/newrelic/dkenv   [no test files]
ok      github.com/newrelic/dkenv/cli   0.013s  coverage: 89.1% of statements
ok      github.com/newrelic/dkenv/lib   0.015s  coverage: 0.9% of statements
```

Commit:
updated readme

idleyoungman committed:
Add minimal .travis.yml
